### ISSUE TITLE
Csc Task Issue with CTP6

### ISSUE THREAD

**Issue by christophwille** [dotnet/roslyn#813]
https://github.com/icsharpcode/NRefactory/tree/roslyn

This (NRefactory.sln) builds with CTP5, the Preview Build Tools as well as Mono mcs. With CTP6, the Tests projects errors out with a single line error message:

C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.CSharp.CurrentVersion.targets(383,9): error MSB6006: "csc2.exe" exited with code 13838984.

(the line in question is the Csc target)

Is there a way to dig deeper into this error message?

**Comment by agocke** [comment]
That usually means the compiler is crashing. Try starting up VBCSCompiler in a debugger or set the property:

`<CscToolExe>csc.exe</CscToolExe>`

**Comment by agocke** [comment]
Someone got bit by #271 with CTP6 -- you should see if that's also your problem.

**Comment by christophwille** [comment]
I added CscToolExe to the project - I now get the following output:

1>------ Build started: Project: ICSharpCode.NRefactory6.Tests, Configuration: Debug Any CPU ------
1>  All packages listed in packages.config are already installed.
1>CSC : warning CS8032: An instance of analyzer Microsoft.CodeAnalysis.CSharp.Analyzers.MetaAnalyzers.CSharpReportDiagnosticAnalyzer cannot be created from D:\GitWorkspace\NRefactory\packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers\C#\Microsoft.CodeAnalysis.CSharp.Analyzers.dll : Could not load file or assembly 'Microsoft.CodeAnalysis.Analyzers, Version=1.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Microsoft.CodeAnalysis.CSharp.Analyzers.MetaAnalyzers.CSharpDiagnosticDescriptorCreationAnalyzer cannot be created from D:\GitWorkspace\NRefactory\packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers\C#\Microsoft.CodeAnalysis.CSharp.Analyzers.dll : Could not find file 'D:\GitWorkspace\NRefactory\packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers\C#\Microsoft.CodeAnalysis.Analyzers.dll'..
1>CSC : warning CS8032: An instance of analyzer Microsoft.CodeAnalysis.CSharp.Analyzers.MetaAnalyzers.CSharpDiagnosticAnalyzerFieldsAnalyzer cannot be created from D:\GitWorkspace\NRefactory\packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers\C#\Microsoft.CodeAnalysis.CSharp.Analyzers.dll : Could not find file 'D:\GitWorkspace\NRefactory\packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers\C#\Microsoft.CodeAnalysis.Analyzers.dll'..
1>CSC : warning CS8032: An instance of analyzer Microsoft.CodeAnalysis.CSharp.Analyzers.MetaAnalyzers.CSharpRegisterActionAnalyzer cannot be created from D:\GitWorkspace\NRefactory\packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers\C#\Microsoft.CodeAnalysis.CSharp.Analyzers.dll : Could not find file 'D:\GitWorkspace\NRefactory\packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers\C#\Microsoft.CodeAnalysis.Analyzers.dll'..
1>C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.CSharp.CurrentVersion.targets(383,9): error MSB6006: "csc.exe" exited with code 255.
========== Build: 0 succeeded, 1 failed, 2 up-to-date, 0 skipped ==========

**Comment by agocke** [comment]
Certainly looks similar to #271. If you debug the compiler while it's failing I'll bet you get a similar exception.

**Comment by christophwille** [comment]
Mind to give me a pointer for the proper debugging? (bonus question: workaround in CTP6?)

**Comment by tmeschter** [comment]
@mavasani: Does this look like issue #271 to you? If so, can you post the workaround?

**Comment by agocke** [comment]
My strategy for debugging is usually to find VBCSCompiler (in your NuGet package if you import Microsoft.Net.Compilers or Microsoft.Net.ToolsetCompilers) and start it in a debugger. csc2.exe will attempt to use a VBCSCompiler that's already been started, so it should reuse the one you started. Then if VBCSCompiler crashes you can watch it in the debugger.

If it looks like csc2.exe is starting a new VBCSCompiler, you are probably running vbcscompiler and csc2 with different permissions.

**Comment by tmeschter** [comment]
One workaround would be to remove the Microsoft.CodeAnalysis.Analyzers NuGet from your projects.

**Comment by christophwille** [comment]
@tmeschter in our specific project (NRefactory), removing those isn't exactly an option as we build on top of it.

**Comment by agocke** [comment]
@christophwille Couldn't you include them as a reference but not an analyzer?

**Comment by christophwille** [comment]
I removed Microsoft.CodeAnalysis.Analyzers from the Analyzers node (kept Microsoft.CodeAnalysis.CSharp.Analyzers). Also removed the CscToolExe (so I am back to csc2.exe).

It builds now, but the warnings remained (the ones that I posted previously). But the "it is building now" part is the good news.

**Comment by agocke** [comment]
@christophwille @mavasani This is almost definitely #271.

**Comment by agocke** [comment]
As a broader issue, we need to make sure that exceptions can't bubble out of analyzers and crash the compiler.

**Comment by tmeschter** [comment]
@christophwille The Microsoft.CodeAnalysis.Analyzers NuGet package contains no reference assemblies, only analyzers. You should be able to remove it entirely without breaking your build--you just won't have access to those analyzers.

**Comment by christophwille** [comment]
@tmeschter tried to remove it, but other packages depend on it:

"Unable to uninstall 'Microsoft.CodeAnalysis.Analyzers 1.0.0-rc1' because 'Microsoft.CodeAnalysis.Common 1.0.0-rc1' depends on it."

**Comment by agocke** [comment]
@srivatsn Do the compiler references somehow require these analyzers? I can understand tying the compiler tools packages to new analyzer versions, but why references?

**Comment by mavasani** [comment]
@tmeschter @christophwille Workaround is on the analyzer author side, I am planning to make the change on the CodeAnalyzers analyzer assembly. Only workaround from user side is to remove the analyzer reference.

**Comment by mavasani** [comment]
@tmeschter @agocke @christophwille Actually, I don't think this is #271 
That is related to unable to find the resources assembly "'Microsoft.CodeAnalysis.Analyzers.resources". The call stack on this issue shows that the analyzer assembly itself "'Microsoft.CodeAnalysis.Analyzers" could not be found.

Can you verify if this assembly lies in the folder where C# analyzer assembly "Microsoft.CodeAnalysis.CSharp.Analyzers.dll" is installed, i,e. D:\GitWorkspace\NRefactory\packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers\C#\

My guess is the nuget package is installing the assembly in a the parent folder, common to C# and VB analyzers. If so, you can workaround this by copying Microsoft.CodeAnalysis.Analyzers.dll to C# and VB folders.

**Comment by agocke** [comment]
@mavasani I think it's the same class of issues, namely that the analyzers engine can't handle missing references without throwing an exception.

**Comment by srivatsn** [comment]
@agocke - the compiler references don't depend on the analyzers. The codeanalysis package declares the analyzers packages as a dependency so that the analyzers get pulled in when someone installs the codeanalysis package. Nuget's working on the notion of an optional dependency and when they have that this will be made optional so that folks can uninstall the analyzers if they choose to.

**Comment by srivatsn** [comment]
Failing to find the reference shouldn't crash the compiler. When CscToolExe is set, it looks like it simply logs a warning. I wonder why in the VBCSCompiler case, it crashing. @tmeschter  can you please take a look at this?

**Comment by agocke** [comment]
@srivatsn I think the compiler is failing to find referenced assemblies, then the analyzer is crashing the compiler. Note the last line in the build log:

    1>C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.CSharp.CurrentVersion.targets(383,9): error 
    MSB6006: "csc.exe" exited with code 255.

**Comment by christophwille** [comment]
I checked the locations:

Packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers for Microsoft.CodeAnalysis.Analyzers.dll

and

Packages\Microsoft.CodeAnalysis.Analyzers.1.0.0-rc1\tools\analyzers\C# for Microsoft.CodeAnalysis.CSharp.Analyzers.dll

But you can always pull our code pre-removal of the analyzer reference and debug directly.

**Comment by mavasani** [comment]
_Update: in the other two projects the analyzer is still referenced and no crashes._

@christophwille Does opening the solution in IDE show you any "RSXXXX" live diagnostics in the error list for the project where command line build crashes? Above update gives me a sense that the crash is coming while trying to report a diagnostic generated by this analyzer, and failing to find the resources (essentially #271).

**Comment by christophwille** [comment]
Yes, I get one RS warning:

Warning	RS1001	Missing 'Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzerAttribute' attribute.	ICSharpCode.NRefactory6.Tests	D:\GitWorkspace\NRefactory\ICSharpCode.NRefactory.Tests\CSharp\CodeIssues\InspectionActionTestBase.cs	209

This is the following:

https://github.com/icsharpcode/NRefactory/blob/roslyn/ICSharpCode.NRefactory.Tests/CSharp/CodeIssues/InspectionActionTestBase.cs#L209

**Comment by mavasani** [comment]
Alright this is #271 I believe. I am going to check in a workaround fix into CodeAnalysis to stop crashing for this scenario in command line builds. I'll continue investigating the root cause of being unable to find the resources from main assembly during command line builds.

**Comment by mavasani** [comment]
@christophwille By the way, if you fix/suppress that RS diagnostic as appropriate, the crash probably goes away. Can you give that a try?

**Comment by tmeschter** [comment]
I've reproduced the crash locally, and I agree that this is a dupe of #271.

There are also two other issues I see here.

  1. The CS8032 warnings you get are because Microsoft.CodeAnalysis.Analyzers.dll is not located next to Microsoft.CodeAnalysis.CSharp.Analyzers.dll, which is where we expect to find it. I filed #828.
  2. In the course of investigating this, I also discovered that the "Compiler Platform Console Application" template does not actually include the Microsoft.CodeAnalysis.Analyzers package. This is #829.
